src/bpfprogram.cpp

```diff
@@ -0,0 +1,74 @@
+#include "bpfprogram.h"
+
+#include <optional>
+#include <stdexcept>
+#include <tuple>
+
+namespace bpftrace {
+
+std::optional<BpfProgram> BpfProgram::CreateFromBytecode(
+    const BpfBytecode &bytecode,
+    const std::string &name,
+    MapManager &maps)
+{
+  if (bytecode.find(name) != bytecode.end())
+  {
+    return BpfProgram(bytecode, name, maps);
+  }
+  return std::nullopt;
+}
+
+BpfProgram::BpfProgram(const BpfBytecode &bytecode,
+                       const std::string &name,
+                       MapManager &maps)
+    : bytecode_(bytecode), maps_(maps), name_(name), code_()
+{
+}
+
+const std::vector<uint8_t> &BpfProgram::getCode()
+{
+  return code_;
+}
+
+void BpfProgram::assemble()
+{
+  if (!code_.empty())
+    return;
+
+  auto &section = bytecode_.at(name_);
+  code_.reserve(section.size());
+  code_.assign(section.data(), section.data() + section.size());
```

Comment:
```suggestion
  auto handle = bytecode_.extract(name_);
  code_ = std::move(handle.mapped());
```
To avoid copying the bytecode unnecessarily

Reply:
I'm a bit skeptical about this one, for two reasons:

1. It'd require us to store `BpfBytecode` as a non-const reference in `BpfProgram`, allowing accidental unwanted modifications.
2. There exists a 1-to-N relationship between `BpfBytecode` (essentially representing the entire ELF created by codegen) and `BpfProgram` (representing the instructions attached to a single probe). While it'd be possible to move out the section with probe instructions from `BpfBytecode` to `BpfProgram`, as it's not supposed to be used in another `BpfProgram`, this won't be possible with other sections that we intend to add in near future (`.BTF`, `.BTF.ext`, `.text`) b/c those will be re-used among multiple `BpfProgram`s.

With all this, I'm inclined towards keeping the copy here for the sake of avoiding hard-to-find bugs where one probe messes with the code of another.

Reply:
Good points. How about we just make `code_` be a reference into `bytecode_` instead of copying it? It doesn't look like it's modified anywhere

Reply:
Good points. Can we just make `code_` a reference into `bytecode_` instead? Then we avoid the copy and the data ownership is essentially the same as what we had before.

The only modification we make to `code_` is relocation, which I think is safe to do multiple times (not that we would currently do this anyway).

Reply:
ATM yes, but that will change once we add support for subprograms. Most importantly, we'll need to append `.text` to the current `code_` and pass the entire thing to libbpf. So, I don't think that we'll avoid copying it but since I'm adding the support in parts, I can just make it a reference for now.

Reply:
And we'd still need to make the reference to `BpfBytecode` non-const.

Reply:
Ah yes, I see it now looking at the full stack. Fair enough, if it'll make the whole thing simpler then let's just always copy it here for now.

I'm not sure what the purpose of the reserve and assign originally were, can they not be simplified to this?
```suggestion
  code_ = bytecode_.at(name_);
```

Reply:
It will make the whole thing much simpler for the moment. Also, I'm working on some follow-ups and it's possible that we'll need to introduce more abstraction to `BpfProgram` so I may revisit this later.

> I'm not sure what the purpose of the reserve and assign originally were, can they not be simplified to this?

I don't see a reason why not, I updated it.